internal/nas/nas_security/security.go

```diff
@@ -168,48 +155,189 @@ func Decode(ue *context.AmfUe, accessType models.AccessType, payload []byte) (*n
-			ue.NASLog.Tracef("cmac value: 0x%08x", mac32)
-			ue.MacFailed = false
+			ue.NASLog.Debugln("UE Security Context is not Available, so skip MAC verify")
 		}
 
 		if ciphered {
-			ue.NASLog.Debugf("Decrypt NAS message (algorithm: %+v, ULCount: 0x%0x)", ue.CipheringAlg, ue.ULCount.Get())
+			if !integrityProtected {
+				return nil, false, fmt.Errorf("NAS message is ciphered, but MAC verification failed")
+			}
+			ue.NASLog.Debugf("Decrypt NAS message (algorithm: %+v, ULCount: 0x%0x)", ue.CipheringAlg, ulCountNew.Get())
 			ue.NASLog.Tracef("NAS ciphering key: %0x", ue.KnasEnc)
 			// decrypt payload without sequence number (payload[1])
-			if err = security.NASEncrypt(ue.CipheringAlg, ue.KnasEnc, ue.ULCount.Get(), GetBearerType(accessType),
+			if err = security.NASEncrypt(ue.CipheringAlg, ue.KnasEnc, ulCountNew.Get(), GetBearerType(accessType),
 				security.DirectionUplink, payload[1:]); err != nil {
-				return nil, fmt.Errorf("Encrypt error: %+v", err)
+				return nil, false, fmt.Errorf("decrypt error: %+v", err)
 			}
 		}
 
 		// remove sequece Number
 		payload = payload[1:]
-		err = msg.PlainNasDecode(&payload)
-		return msg, err
 	}
+
+	err = msg.PlainNasDecode(&payload)
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	msgTypeText := func() string {
+		if msg.GmmMessage == nil {
+			return "Non GMM message"
+		} else {
+			return fmt.Sprintf(" message type %d", msg.GmmHeader.GetMessageType())
+		}
+	}
+	errNoSecurityContext := func() error {
+		return fmt.Errorf("UE Security Context is not Available, %s", msgTypeText())
+	}
+	errWrongSecurityHeader := func() error {
+		return fmt.Errorf("wrong security header type: 0x%0x, %s", msg.SecurityHeader.SecurityHeaderType, msgTypeText())
+	}
+	errMacVerificationFailed := func() error {
+		return fmt.Errorf("MAC verification failed, %s", msgTypeText())
+	}
+
+	if msg.GmmMessage == nil {
+		if !ue.SecurityContextAvailable {
+			return nil, false, errNoSecurityContext()
+		}
+		if msg.SecurityHeaderType != nas.SecurityHeaderTypeIntegrityProtectedAndCiphered {
+			return nil, false, errWrongSecurityHeader()
+		}
+		if !integrityProtected {
+			return nil, false, errMacVerificationFailed()
+		}
+	} else {
+		switch msg.GmmHeader.GetMessageType() {
+		case nas.MsgTypeDeregistrationRequestUEOriginatingDeregistration, nas.MsgTypeRegistrationRequest:
+			if initialMessage {
+				if msg.SecurityHeaderType == nas.SecurityHeaderTypeIntegrityProtectedAndCiphered ||
+					msg.SecurityHeaderType == nas.SecurityHeaderTypeIntegrityProtectedAndCipheredWithNew5gNasSecurityContext {
+					return nil, false, errWrongSecurityHeader()
+				}
+			} else {
+				if ue.SecurityContextAvailable {
+					if msg.SecurityHeaderType != nas.SecurityHeaderTypeIntegrityProtectedAndCiphered {
+						return nil, false, errWrongSecurityHeader()
+					}
+					if !integrityProtected {
+						return nil, false, errMacVerificationFailed()
+					}
+				}
+			}
+		case nas.MsgTypeServiceRequest:
+			if initialMessage {
```

Comment:
Hi @ShouheiNishi 
Refer to 4.4.5 Ciphering of NAS signalling messages from 3GPP TS 24.501:
>  The AMF shall start ciphering and deciphering of NAS messages as described in subclause 4.4.2.5. From this time onward, except for the SECURITY MODE COMMAND message, the AMF shall send all NAS messages ciphered until the N1 NAS signalling connection is released, or the UE performs inter-system change to S1 mode. 

I think we don't need the case of initialMessage, because of the UEContextReleaseComplete means n2 release not n1 release. The n1 release happens when AMF receives the dereg request.

@tim-ywliu  @brianchennn FYI